src/asynqp/routing.py

```diff
@@ -56,54 +57,48 @@ def handle(self, frame):
         meth(frame)
 
     def handle_PoisonPillFrame(self, frame):
-        self.synchroniser.killall(ConnectionError)
+        self.synchroniser.killall(frame.exception)
 
 
 class Synchroniser(object):
-    _blocking_methods = set((spec.BasicCancelOK,  # Consumer.cancel
-                             spec.ChannelCloseOK,  # Channel.close
-                             spec.ConnectionCloseOK))  # Connection.close
 
     def __init__(self, *, loop):
         self._loop = loop
-        self._futures = OrderedManyToManyMap()
-        self.connection_closed = False
+        self._futures = collections.defaultdict(collections.deque)
+        self.connection_exc = None
 
     def await(self, *expected_methods):
         fut = asyncio.Future(loop=self._loop)
 
-        if self.connection_closed:
-            for method in expected_methods:
-                if method in self._blocking_methods and not fut.done():
-                    fut.set_result(None)
-            if not fut.done():
-                fut.set_exception(ConnectionError)
+        if self.connection_exc is not None:
+            fut.set_exception(self.connection_exc)
             return fut
 
-        self._futures.add_item(expected_methods, fut)
+        for method in expected_methods:
+            self._futures[method].append((fut, expected_methods))
         return fut
 
     def notify(self, method, result=None):
-        fut = self._futures.get_leftmost(method)
-        fut.set_result(result)
-        self._futures.remove_item(fut)
+        try:
+            fut, bound_methods = self._futures[method].popleft()
+        except IndexError:
+            # XXX: we can't just ignore this.
+            log.error("Got an unexpected method notification %s", method)
+        else:
+            # Cleanup futures, that were awaited together, like
+            # (spec.BasicGetOK, spec.BasicGetEmpty)
+            for other_method in bound_methods:
+                if other_method != method:
+                    self._futures[other_method].remove((fut, bound_methods))
```

Comment:
This is expensive. `deque.remove` has to search through the deque for the item it's looking for, and removing it means copying a potentially large number of other items into place. In other words `remove` is O(n), whereas it was amortised O(1) when it was based on `OrderedDict` before